### DISCUSSION TITLE
Using alternative compilation of SQLite doesn't work

### DISCUSSION THREAD

**Discussion by fractaledmind** [sparklemotion/sqlite3-ruby#400] [Q&A]
I am trying to get a project-local, custom build of SQLite working with my Rails app. I have been studying and using the [installation instructions](https://github.com/sparklemotion/sqlite3-ruby/blob/master/INSTALLATION.md#avoiding-the-precompiled-native-gem), but I can't quite get things to work and I don't know what I'm doing wrong. I'm hoping someone here can help steer me in the right direction.

I have the following script which compiles a custom, optimized version of SQLite:
```shell
#!/usr/bin/env sh

cd ../vendor
mkdir -p sqlite/bin
mkdir -p sqlite/lib
mkdir -p sqlite/include

# ============================================================================================================
# Compile and install sqlite3 (for performance turning and build customizations)
# SEE: https://www.sqlite.org/compile.html
# NOTE: The sqlite3 Ruby gem will not work with the following compile time flags
#       * -DSQLITE_OMIT_DEPRECATED
# ============================================================================================================
curl --remote-name --remote-header-name https://www.sqlite.org/2023/sqlite-amalgamation-3430000.zip && unzip sqlite-amalgamation-3430000.zip

cd sqlite-amalgamation-3430000

PREPROCESSOR_FLAGS=(
  -DSQLITE_DEFAULT_MEMSTATUS=0
  -DSQLITE_DEFAULT_WAL_SYNCHRONOUS=1
  -DSQLITE_DQS=0
  -DSQLITE_ENABLE_FTS5
  -DSQLITE_LIKE_DOESNT_MATCH_BLOBS
  -DSQLITE_MAX_EXPR_DEPTH=0
  -DSQLITE_OMIT_PROGRESS_CALLBACK
  -DSQLITE_OMIT_SHARED_CACHE
  -DSQLITE_USE_ALLOCA
)

# compile the executable
gcc "${PREPROCESSOR_FLAGS[@]}" shell.c sqlite3.c -lpthread -ldl -lm -o ../sqlite/bin/sqlite3

# compile and setup shared library
gcc "${PREPROCESSOR_FLAGS[@]}" shell.c sqlite3.c -lpthread -ldl -lm -fPIC -shared -o ../sqlite/lib/libsqlite3.so && \
chmod 755 ../sqlite/lib/libsqlite3.so && \
ln -s ../sqlite/lib/libsqlite3.so ../sqlite/lib/libsqlite3.so.0

# copy header/include files
mkdir -p ../sqlite/include/sqlite3 && \
cp sqlite3.h ../sqlite/include/ && \
cp sqlite3ext.h ../sqlite/include/

# cleanup
PREPROCESSOR_FLAGS=""
```

This creates a `vendor/sqlite` directory in my Rails app with this structure:
```
sqlite
├── bin
│   └── sqlite3
├── include
│   ├── sqlite3.h
│   └── sqlite3ext.h
└── lib
    ├── libsqlite3.so
    └── libsqlite3.so.0 -> ../sqlite/lib/libsqlite3.so

4 directories, 5 files
```

I setup the Bundle config like so:
```
bundle config set build.sqlite3 --enable-system-libraries --with-opt-dir=./vendor/sqlite
```
which creates a `.bundle/config` file like so:
```yaml
---
BUNDLE_BUILD__SQLITE3: "--enable-system-libraries --with-opt-dir=./vendor/sqlite"
```

In my Gemfile, I specify the dependency on SQLite like so:
```ruby
gem "sqlite3", force_ruby_platform: true
```

However, whenever I use `bin/rails c` to start a console and run `ActiveRecord::Base.connection.execute 'PRAGMA compile_options'`, I see the compile_options of my system SQLite installation at `/usr/bin/sqlite3`. 

When I manually enter the compiled SQLite prompt via `vendor/sqlite/bin/sqlite3` and run `PRAGMA compile_options`, I see the specific compile options that I set.

- - -

So, how can I get the SQLite installation that lives in `vendor/sqlite` to be the one that the gem uses in my Rails app?

**Comment by flavorjones** [comment]
@fractaledmind I wrote a bit about this topic in this discussion: https://github.com/sparklemotion/sqlite3-ruby/discussions/389

Specifically these comments:
- https://github.com/sparklemotion/sqlite3-ruby/discussions/389#discussioncomment-5967800
- https://github.com/sparklemotion/sqlite3-ruby/discussions/389#discussioncomment-5970885

Can you take a look and see if that's helpful? If so I should probably add an explanation to the README.

**Reply by flavorjones** [reply]
Ah, I already wrote it up in a markdown file!

https://github.com/sparklemotion/sqlite3-ruby/blob/master/INSTALLATION.md#how-do-i-use-an-alternative-sqlite3-implementation

That presumes you've downloaded the autoconf source package from https://www.sqlite.org/download.html and have built it using `./configure --prefix=/opt/whatever` along with whatever compile-time options you need.

**Reply by fractaledmind** [reply]
This was the section I was following. I believe I have followed the details in that section of the `INSTALLATION.md` doc. The issue is that when I introspect my database from my Rails console via `PRAGMA compile_options`, the compile options I used to compile my SQLite aren't present. Calling that PRAGMA from within the sqlite3 terminal started by pointed at the executable I compiled in `vendor/sqlite/bin` does show the options I set, which leads me to believe that the gem is using a different SQLite executable

**Comment by flavorjones** [comment]
I'm realizing now that you may be asking about a different thing: how do you get your system to prefer loading this library from your custom sqlite instead of the system's?

Try setting the environment variable `LD_LIBRARY_PATH=/path/to/rails/app/vendor/sqlite/lib`

**Reply by fractaledmind** [reply]
Perhaps. I don't want my entire system to use this project-specific SQLite instance, only my Rails app as wired up via the ruby-sqlite3 gem.

**Reply by flavorjones** [reply]
you don't need to make this a system-wide setting; you should just set the environment variable for the ruby process that is running the Rails app.

**Reply by fractaledmind** [reply]
That seems to work. But I am now experiencing a new issue. I will open a new discussion for it tho to keep things clear.

**Reply by flavorjones** [reply]
I'll add a note to the INSTALLATION.md doc about LD_LIBRARY_PATH. Thanks for closing the loop.

**Reply by fractaledmind** [reply]
Actually, I think I spoke too soon. Sorry.

**Reply by fractaledmind** [reply]
I wish there were an easier way to check where the gem is loading SQLite from, for now I can only fingerprint via `PRAGMA compile_options;` and checking the "fingerprint" of that output. I am seeing three different fingerprints, which confuses me. I can see the fingerprint for a new Rails app with SQLite; this makes sense. I can see the fingerprint for the custom compiled SQLite executable in my vendor directory; and this makes sense. But the fingerprint I see in my Rails console in the app where I installed the sqlite3 gem using the custom installation instructions isn't either of the two other fingerprints and it makes no sense to me.

Here are the "fingerprints":

The fingerprint for a new, basic Rails app with SQLite is:
```
ATOMIC_INTRINSICS=1
COMPILER=clang-10.0.0
DEFAULT_AUTOVACUUM
DEFAULT_CACHE_SIZE=-2000
DEFAULT_FILE_FORMAT=4
DEFAULT_JOURNAL_SIZE_LIMIT=-1
DEFAULT_MMAP_SIZE=0
DEFAULT_PAGE_SIZE=4096
DEFAULT_PCACHE_INITSZ=20
DEFAULT_RECURSIVE_TRIGGERS
DEFAULT_SECTOR_SIZE=4096
DEFAULT_SYNCHRONOUS=2
DEFAULT_WAL_AUTOCHECKPOINT=1000
DEFAULT_WAL_SYNCHRONOUS=2
DEFAULT_WORKER_THREADS=0
ENABLE_FTS3
ENABLE_FTS4
ENABLE_FTS5
ENABLE_GEOPOLY
ENABLE_MATH_FUNCTIONS
ENABLE_RTREE
MALLOC_SOFT_LIMIT=1024
MAX_ATTACHED=10
MAX_COLUMN=2000
MAX_COMPOUND_SELECT=500
MAX_DEFAULT_PAGE_SIZE=8192
MAX_EXPR_DEPTH=1000
MAX_FUNCTION_ARG=127
MAX_LENGTH=1000000000
MAX_LIKE_PATTERN_LENGTH=50000
MAX_MMAP_SIZE=0x7fff0000
MAX_PAGE_COUNT=1073741823
MAX_PAGE_SIZE=65536
MAX_SQL_LENGTH=1000000000
MAX_TRIGGER_DEPTH=1000
MAX_VARIABLE_NUMBER=32766
MAX_VDBE_OP=250000000
MAX_WORKER_THREADS=8
MUTEX_PTHREADS
SYSTEM_MALLOC
TEMP_STORE=1
THREADSAFE=1
 ```
 
 The fingerprint for my custom compiled SQLite executable in `vendor/sqlite/bin/sqlite3` is:
 ```
 ATOMIC_INTRINSICS=1
COMPILER=clang-14.0.0
DEFAULT_AUTOVACUUM
DEFAULT_CACHE_SIZE=-2000
DEFAULT_FILE_FORMAT=4
DEFAULT_JOURNAL_SIZE_LIMIT=-1
DEFAULT_MEMSTATUS=0
DEFAULT_MMAP_SIZE=0
DEFAULT_PAGE_SIZE=16384
DEFAULT_PCACHE_INITSZ=20
DEFAULT_RECURSIVE_TRIGGERS
DEFAULT_SECTOR_SIZE=4096
DEFAULT_SYNCHRONOUS=2
DEFAULT_WAL_AUTOCHECKPOINT=1000
DEFAULT_WAL_SYNCHRONOUS=1
DEFAULT_WORKER_THREADS=0
DQS=0
ENABLE_FTS5
LIKE_DOESNT_MATCH_BLOBS
MALLOC_SOFT_LIMIT=1024
MAX_ATTACHED=10
MAX_COLUMN=2000
MAX_COMPOUND_SELECT=500
MAX_DEFAULT_PAGE_SIZE=8192
MAX_EXPR_DEPTH=0
MAX_FUNCTION_ARG=127
MAX_LENGTH=1000000000
MAX_LIKE_PATTERN_LENGTH=50000
MAX_MMAP_SIZE=0x7fff0000
MAX_PAGE_COUNT=1073741823
MAX_PAGE_SIZE=65536
MAX_SQL_LENGTH=1000000000
MAX_TRIGGER_DEPTH=1000
MAX_VARIABLE_NUMBER=32766
MAX_VDBE_OP=250000000
MAX_WORKER_THREADS=0
MUTEX_OMIT
OMIT_DECLTYPE
OMIT_PROGRESS_CALLBACK
OMIT_SHARED_CACHE
SYSTEM_MALLOC
TEMP_STORE=1
THREADSAFE=0
USE_ALLOCA
```

And then, the fingerprint from within the Rails console, using the `LD_LIBRARY_PATH` (tho this doesn't seem to make a difference actually), is:
```

ATOMIC_INTRINSICS=1
BUG_COMPATIBLE_20160819
COMPILER=clang-13.1.6
DEFAULT_AUTOVACUUM
DEFAULT_CACHE_SIZE=2000
DEFAULT_CKPTFULLFSYNC
DEFAULT_FILE_FORMAT=4
DEFAULT_JOURNAL_SIZE_LIMIT=32768
DEFAULT_LOOKASIDE=1200,102
DEFAULT_MEMSTATUS=0
DEFAULT_MMAP_SIZE=0
DEFAULT_PAGE_SIZE=4096
DEFAULT_PCACHE_INITSZ=20
DEFAULT_RECURSIVE_TRIGGERS
DEFAULT_SECTOR_SIZE=4096
DEFAULT_SYNCHRONOUS=2
DEFAULT_WAL_AUTOCHECKPOINT=1000
DEFAULT_WAL_SYNCHRONOUS=1
DEFAULT_WORKER_THREADS=0
ENABLE_API_ARMOR
ENABLE_BYTECODE_VTAB
ENABLE_COLUMN_METADATA
ENABLE_DBSTAT_VTAB
ENABLE_FTS3
ENABLE_FTS3_PARENTHESIS
ENABLE_FTS3_TOKENIZER
ENABLE_FTS4
ENABLE_FTS5
ENABLE_JSON1
ENABLE_LOCKING_STYLE=1
ENABLE_NORMALIZE
ENABLE_PREUPDATE_HOOK
ENABLE_RTREE
ENABLE_SESSION
ENABLE_SNAPSHOT
ENABLE_SQLLOG
ENABLE_STMT_SCANSTATUS
ENABLE_UNKNOWN_SQL_FUNCTION
ENABLE_UPDATE_DELETE_LIMIT
HAS_CODEC_RESTRICTED
HAVE_ISNAN
MALLOC_SOFT_LIMIT=1024
MAX_ATTACHED=10
MAX_COLUMN=2000
MAX_COMPOUND_SELECT=500
MAX_DEFAULT_PAGE_SIZE=8192
MAX_EXPR_DEPTH=1000
MAX_FUNCTION_ARG=127
MAX_LENGTH=2147483645
MAX_LIKE_PATTERN_LENGTH=50000
MAX_MMAP_SIZE=1073741824
MAX_PAGE_COUNT=1073741823
MAX_PAGE_SIZE=65536
MAX_SQL_LENGTH=1000000000
MAX_TRIGGER_DEPTH=1000
MAX_VARIABLE_NUMBER=500000
MAX_VDBE_OP=250000000
MAX_WORKER_THREADS=8
MUTEX_UNFAIR
OMIT_AUTORESET
OMIT_LOAD_EXTENSION
STMTJRNL_SPILL=131072
SYSTEM_MALLOC
TEMP_STORE=1
THREADSAFE=2
USE_URI
```

As you can see, the third set of compile options is quite unique from the other two. I don't know, for example, where the `OMIT_LOAD_EXTENSION` option is coming from (this is unlike the default), why `MAX_EXPR_DEPTH` is still `1000` (this is unlike my custom compilation), and why the `COMPILER` is a completely different clang version than either of the other two. 

I just want my Rails console to use my custom SQLite, but I still haven't figured out how to do that, and I appear to have created some bizarre third SQLite instance that Rails is using for some reason.

**Reply by flavorjones** [reply]
One way to see what shared libraries are being loaded on Linux is to `cat /proc/#{PID}/maps` and grep for the sqlite library.

**Reply by flavorjones** [reply]
I have some time today if you want to pair on this -- I DMed you on twitter.

**Reply by flavorjones** [reply]
Stepping back for a minute ... it seems like the key thing you want to do is to compile sqlite3 with certain options.

Have you tried setting `CFLAGS` at gem install time?

```sh
$ CFLAGS=-DSQLITE_DEFAULT_PAGE_SIZE=16300 gem install sqlite3 --platform=ruby
Fetching sqlite3-1.6.4.gem
Building native extensions. This could take a while...
Successfully installed sqlite3-1.6.4
1 gem installed

$ ruby -rsqlite3 -e 'puts SQLite3::Database.new(":memory:").execute("pragma page_size")'
16300
```

This doesn't require any changes to the gem extconf.rb, does that help you do what you want?

**Reply by fractaledmind** [reply]
Let me try. You are write, all I want to do is set compile options for SQLite. Is there any way to do this thru bundler such that this can be integrated into an app repo and not system level?

**Reply by flavorjones** [reply]
Well, you can ensure you get the source gem (not a precompiled gem) with:

```ruby
# Gemfile
gem "sqlite3", force_ruby_platform: true # needs bundler 2.3.18 or higher
```

But the CFLAGS environment variable you'll have to set outside of bundler and rails. If you're building a docker image, I'd recommend doing it in the Dockerfile. Or else consider using `direnv` or something else that lets you use a `.envrc` file in the repository? Be careful with how you build your production images, though, that those environment variables take effect.

**Reply by flavorjones** [reply]
Thinking :thinking: We could add an extconf option to allow people to pass the cflags via the bundler config ... that would hit your use case. Something like

```sh
bundle config build.nokogiri --sqlite-cflags="-DSQLITE_DEFAULT_PAGE_SIZE=16300"
```

WDYT?

**Reply by fractaledmind** [reply]
Yea, I was trying to find the gem that I have a fuzzy memory of that allowed compilation options passed via a bundled config. But my dizzy memory looked like that. 

FWIW, this discussion does show a clear benefit of Docker. I just want to try and find something that would work with a Hatchbox deployment, for example, without Docker. Because I think that opens things up to everyone. 

So, short answer, if we could get the gem to support a builder config like the above, I think we would unlock a really cool feature (fine tuning your db for your app) for every Rails app regardless of deployment strategy, which would be insanely cool IMO

**Reply by flavorjones** [reply]
:+1: see https://github.com/sparklemotion/sqlite3-ruby/issues/401

**Comment by flavorjones** [comment]
Closing the loop on this, @fractaledmind and I spec'ed out the DX to allow developers to set compiler flags at gem-install time in a bundler config. See [Release 1.6.5 / 2023-09-08 · sparklemotion/sqlite3-ruby](https://github.com/sparklemotion/sqlite3-ruby/releases/tag/v1.6.5) for details.
